jsonrpc2/backoff_handler.go

```diff
@@ -0,0 +1,154 @@
+}
+
+type backoffTimer struct {
+	retries           float64
+	lastAttemptedTime *time.Time
+	lastDurationTime  time.Duration
+}
+
+func (b *backoffTimer) BackoffRequest() time.Duration {
+	if b.lastAttemptedTime == nil {
+		t := time.Now()
+		b.lastAttemptedTime = &t
+		b.lastDurationTime = time.Duration(0)
+		return b.lastDurationTime
+	}
+
+	// if backoff exists but has been more than a minute passed the
+	// the last back off, then reset the backoff.
+	if time.Now().After(b.lastAttemptedTime.Add(b.lastDurationTime).Add(time.Minute * 1)) {
+		b.retries = 0
+		t := time.Now()
+		b.lastAttemptedTime = &t
+		b.lastDurationTime = time.Duration(0)
+		return b.lastDurationTime
+	}
+
+	// calculated back off
+	b.lastDurationTime = time.Second * time.Duration((math.Pow(2, b.retries)))
+
+	// Cap back off at 5 min.
+	if b.lastDurationTime >= (time.Minute * 5) {
+		b.lastDurationTime = time.Minute * 5
+	}
+
+	b.retries = b.retries + 1
+	return b.lastDurationTime
+}
+
+var _ Handler = &BackoffHandler{}
+
+// Cancel is invoked for cancelled outgoing requests.
+// It is okay to use the connection to send notifications, but the context will
+// be in the cancelled state, so you must do it with the background context
+// instead.
+// If Cancel returns true all subsequent handlers will be invoked with
+// cancelled set to true, and should not attempt to cancel the message.
+func (b *BackoffHandler) Cancel(ctx context.Context, conn *Conn, id ID, cancelled bool) bool {
+	return cancelled
+}
+
+// Log is invoked for all messages flowing through a Conn.
+// direction indicates if the message being received or sent
+// id is the message id, if not set it was a notification
+// elapsed is the time between a call being seen and the response, and is
+// negative for anything that is not a response.
+// method is the method name specified in the message
+// payload is the parameters for a call or notification, and the result for a
+// response
+// Request is called near the start of processing any request.
+func (b *BackoffHandler) Request(ctx context.Context, conn *Conn, direction Direction, r *WireRequest) context.Context {
+	//handle Back off
+	requestKey := requestKey{
+		method: r.Method,
+		params: string(*r.Params),
+	}
+	b.failedRequestsMu.Lock()
+	backOff, ok := b.failedRequests[requestKey]
+	if !ok {
+		backOff = &backoffTimer{
+			retries:           0,
+			lastAttemptedTime: nil,
+			lastDurationTime:  0,
+		}
+		b.failedRequests[requestKey] = backOff
+	}
+	b.failedRequestsMu.Unlock()
+
+	d := backOff.BackoffRequest()
+	b.logger.V(5).Info("starting backing off request", "method", r.Method, "duration", d)
+	time.Sleep(d)
+	b.logger.V(5).Info("stoping backing off request", "method", r.Method)
+	return context.WithValue(ctx, "back-off-timer", requestKey)
+}
+
+// Response is called near the start of processing any response.
+func (b *BackoffHandler) Response(ctx context.Context, conn *Conn, direction Direction, r *WireResponse) context.Context {
```

Comment:
I'm a little confused by these just returning context

Reply:
And why are we only returning errors because we have to be apart of the interface https://github.com/konveyor/analyzer-lsp/blob/main/jsonrpc2/handler.go#L13

And we learned that there is an updated lsp from go so that we don't have to copy, and it has the same interface so we are able to use this, once we use the correct library.

Reply:
I just looked this up, and apparently, the interface is different: https://pkg.go.dev/go.lsp.dev/jsonrpc2#Handler

It does not allow you to have as much control, I don't think that we want to ripe this out right now and test with this package, I can change the interfaces now and update the copied code if you and @pranavgaikwad feel like it is worth it.